File: C++/myLeetCode/Linked_List/intersection-of-two-linked-lists.cpp
// 面试题 02.07. 链表相交
// https://leetcode.cn/problems/intersection-of-two-linked-lists-lcci/

#include "LinkedListUtils.hpp"

ListNode* getIntersectionNode(ListNode* headA, ListNode* headB) {
    if (headA == nullptr || headB == nullptr) {
        return nullptr;
    }
    
    ListNode* pA = headA;
    ListNode* pB = headB;
    
    while (pA != pB) {
        pA = (pA == nullptr) ? headB : pA->next;
        pB = (pB == nullptr) ? headA : pB->next;
    }
    
    return pA;
}

int main() {
    ListNode* headA = new ListNode(4);
    ListNode* node1 = new ListNode(1);
    ListNode* node8 = new ListNode(8);
    ListNode* node4 = new ListNode(4);
    ListNode* node5 = new ListNode(5);
    headA->next = node1;
    node1->next = node8;
    node8->next = node4;
    node4->next = node5;
    
    ListNode* headB = new ListNode(5);
    ListNode* node0 = new ListNode(0);
    ListNode* node1b = new ListNode(1);
    headB->next = node0;
    node0->next = node1b;
    node1b->next = node8; // Intersection point
    
    ListNode* intersection = getIntersectionNode(headA, headB);
    
    if (intersection != nullptr) {
        std::cout << "Intersection point: " << intersection->val << std::endl;
    } else {
        std::cout << "No intersection point found." << std::endl;
    }
    
    return 0;
}

File: C++/代码随想录/Linked_List/intersection-of-two-linked-lists.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

def getIntersectionNode(headA, headB):
    if not headA or not headB:
        return None

    pointerA, pointerB = headA, headB

    while pointerA != pointerB:
        pointerA = pointerA.next if pointerA else headB
        pointerB = pointerB.next if pointerB else headA

    return pointerA

# 测试用例
# 创建两个链表：A: 1->2->3->4->5 和 B: 6->7->8->3->4->5 (相交于节点3)
node1 = ListNode(1)
node2 = ListNode(2)
node3 = ListNode(3)
node4 = ListNode(4)
node5 = ListNode(5)
node6 = ListNode(6)
node7 = ListNode(7)
node8 = ListNode(8)

node1.next = node2
node2.next = node3
node3.next = node4
node4.next = node5

node6.next = node7
node7.next = node8
node8.next = node3

print(getIntersectionNode(node1, node6).val)  # 应输出3
